### PR TITLE
Add activate/deactivate hooks

### PR DESCRIPTION
**Pull request by joshfrench**
`activate`/`deactivate` hooks have been discussed several times but I haven't seen a PR, so... here's one :)

There are two test failures in `test/activate` and `test/deactivate` that I _think_ are just me not fully understanding the stub system, do you mind taking a look and pointing me in the right direction? Thanks so much!

---
Past discussions:
1. Closes https://github.com/pyenv/pyenv-virtualenv/issues/428
2. https://github.com/pyenv/pyenv-virtualenv/issues/55
3.  https://github.com/pyenv/pyenv/issues/2337

### CODE REVIEW

File: test/activate.bats [hunk 1]
```diff
@@ -16,6 +16,11 @@ setup() {
   unset PYENV_VIRTUAL_ENV_DISABLE_PROMPT
   unset VIRTUAL_ENV_DISABLE_PROMPT
   unset _OLD_VIRTUAL_PS1
+  stub pyenv-hooks "activate : echo"
+}
+
+teardown() {
+  unstub pyenv-hooks
```
Comment (native-api) [inline test/activate.bats:23]: "`unstub failed`" happens if the real set of executions does not equal the execution plan that is specified with `stub`. In Pyenv proper, we added code to the `stub` executable to show the execution plan and log upon failure.
If you don't need to validate the plan, you can just not do `unstub`: all stubs are deleted together with the test directory after each test.

[the remark this answers is not in the document]
Comment (joshfrench) [reply test/activate.bats]: These tests fail with `pyenv-hooks: command not found` unless it's stubbed.

**Comment by native-api** [comment]
@pyenv/pyenv-core-maintainers and especially @yyuu : any second opinions on the hook invocation placement?

**Review by native-api — APPROVED** [review]

**Review by yyuu — APPROVED** [review]
looks good 👍 

personally i might not apply very much of customization on activate/deactivate to keep shell setup cleaner/simpler, however, i can understand that there are some demands to customize the behavior around them. i’m open to adopt this definitely (esp. with working patch 😉 )

**Pull request merged**